File: linked-list-cycle/linked-list-cycle.go
package linked_list_cycle

import . "leetcode-go/built-in"

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil || head.Next.Next == nil {
		return false
	}
	slow := head.Next
	fast := head.Next.Next

	for fast != slow && fast.Next != nil && slow.Next != nil {
		slow = slow.Next
		fast = fast.Next
		if fast.Next != nil {
			fast = fast.Next
		} else {
			return false
		}
	}
	return fast == slow
}

File: leetcode-cpp/linked-list-cycle/linked-list-cycle.cpp
//
// Created by smile on 2021/3/16.
//
#include <ListNode.h>

class Solution {
public:
    static bool hasCycle(ListNode *head) {

        if (head == nullptr || head->next == nullptr || head->next->next == nullptr) {
            return false;
        }
        ListNode *Fast, *Slow;
        Slow = head->next;
        Fast = head->next->next;

        while (Fast != Slow && Fast->next != nullptr && Slow->next != nullptr) {
            Slow = Slow->next;
            Fast = Fast->next;
            if (Fast->next != nullptr) {
                Fast = Fast->next;
            } else {
                return false;
            }
        }
        return Fast == Slow;

    }

};